### PR TITLE
Add company detail page

### PR DESCRIPTION
**Pull request by BSaunders95**
Resolves APO-31

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/web/companydocuments/service/CompanyService.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+@Service
+public class CompanyService {
+
+    @Autowired
+    private ApiClientService apiClientService;
+
+    @Autowired
+    private CompanyTransformer companyTransformer;
+
+    private static final UriTemplate COMPANY_URI = new UriTemplate("/company/{companyNumber}");
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/web/companydocuments/service/CompanyService.java]: private statics should come before member vars.

File: src/main/java/uk/gov/companieshouse/web/companydocuments/service/CompanyService.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import uk.gov.companieshouse.web.companydocuments.transformer.CompanyTransformer;
+
+@Service
+public class CompanyService {
+
+    @Autowired
+    private ApiClientService apiClientService;
+
+    @Autowired
+    private CompanyTransformer companyTransformer;
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/web/companydocuments/service/CompanyService.java:21]: Can we start to use constructor injection instead of field injection? It's better practice:
https://reflectoring.io/constructor-injection/

**Review by ch-tstrickland — COMMENTED** [review]
Can we start to use constructor injection instead of field injection? It's better practice:
https://reflectoring.io/constructor-injection/

When compared with other ways of injection, it makes code smells more apparent, tests less brittle and removes possible race conditions.

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package uk.gov.companieshouse.web.companydocuments.util;
+
+import java.util.Map;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class EnumerationConstantsMapper {
+
+    private Map<String, String> companyTypes;
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java]: Both fields could be final.

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    private Map<String, String> companyStatuses;
+
+    @Autowired
+    public EnumerationConstantsMapper(YamlResourceMapper yamlResourceMapper) {
+
+        Map<String, Object> enumerationConstants = yamlResourceMapper.fetchYamlFromFile("api-enumerations/constants.yml");
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:17]: This file is missing from the commits.

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+        this.companyTypes = (Map<String, String>) enumerationConstants.get("company_type");
+        this.companyStatuses = (Map<String, String>) enumerationConstants.get("company_status");
+    }
+
+    public String getCompanyType(String enumeration) {
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:23]: Should this complain (throw ex) if enumeration does not exist in companyTypes?

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/YamlResourceMapper.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    public Map<String, Object> fetchYamlFromFile(String filePath) {
+
+        try {
+            return new Yaml().load(
+                    new FileInputStream(new File(filePath)));
+
+        } catch (FileNotFoundException e) {
+
+            throw new RuntimeException(e);
```
Comment (ch-tstrickland) [inline src/main/java/uk/gov/companieshouse/web/companydocuments/util/YamlResourceMapper.java]: A message would be handy. I've looked at the Java source and not every throw of FileNotFound includes a message, so something along the lines of ("Could not load YAML file: " + filePath, e).
Also throwing RuntimeException makes Sonar wince - perhaps IllegalArgumentEx, or IllegalStateEx?

**Review by ch-tstrickland — COMMENTED** [review]
Can we also avoid InjectMocks.
TLDR: Mockito injection fails silently, which is rubbish and harder to debug.
Here's why:
https://tedvinke.wordpress.com/2014/02/13/mockito-why-you-should-not-use-injectmocks-annotation-to-autowire-fields/

File: src/test/java/uk/gov/companieshouse/web/companydocuments/controller/CompanyDetailsControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    @Mock
+    private CompanyService companyService;
+
+    @Mock
+    private CompanyDetail companyDetail;
+
+    @InjectMocks
+    private CompanyDetailsController controller;
+
+    private static final String COMPANY_NUMBER = "companyNumber";
```
Comment (ch-tstrickland) [inline src/test/java/uk/gov/companieshouse/web/companydocuments/controller/CompanyDetailsControllerTest.java]: Constants should come first.

File: src/test/java/uk/gov/companieshouse/web/companydocuments/service/CompanyServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    @Mock
+    private CompanyDetail companyDetail;
+
+    @Mock
+    private CompanyTransformer companyTransformer;
+
+    @InjectMocks
+    private CompanyService companyService;
+
+    private static final String COMPANY_NUMBER = "companyNumber";
```
Comment (ch-tstrickland) [inline src/test/java/uk/gov/companieshouse/web/companydocuments/service/CompanyServiceTest.java]: Constants should come first.

Comment (BSaunders95) [reply src/test/java/uk/gov/companieshouse/web/companydocuments/service/CompanyServiceTest.java]: Moved in [b916061](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/b9160619da9c40be5d6b3314aec0733f77b248a6) 👍

File: src/test/java/uk/gov/companieshouse/web/companydocuments/controller/CompanyDetailsControllerTest.java (hunk 7 above)
Comment (BSaunders95) [reply src/test/java/uk/gov/companieshouse/web/companydocuments/controller/CompanyDetailsControllerTest.java]: Moved in [b916061](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/b9160619da9c40be5d6b3314aec0733f77b248a6) 👍

File: src/main/java/uk/gov/companieshouse/web/companydocuments/service/CompanyService.java (hunk 1 above)
Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/service/CompanyService.java]: Moved in [b916061](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/b9160619da9c40be5d6b3314aec0733f77b248a6) 👍

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java (hunk 4 above)
Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:17]: That's because api-enumerations is a git submodule

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java (hunk 3 above)
Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java]: Addressed in [2740404](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/2740404f1f799d4c7edb2e0e8f0b3c902f0b98b4)

**Comment by BSaunders95** [comment]
Field injection removed - sorry was on autopilot! Addressed in [b512bd3](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/b512bd3b589813159693b838a87da36ac692f223)

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/YamlResourceMapper.java (hunk 6 above)
Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/YamlResourceMapper.java]: Addressed in [042c7af](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/042c7af9255a7fc7fc142371427169525f946764)

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java (hunk 5 above)
Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:23]: You're quite right. I was hoping to avoid it as it makes the code a little messy - especially given that we should never encounter the scenario. That said, a safety guard isn't a bad shout at all. Added in [5696f8a](https://github.com/companieshouse/company-documents.web.ch.gov.uk/pull/10/commits/5696f8a83b72abac4f4b2aa5b454df5b70aeea50)

File: src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java (hunk 4 above)
Comment (ch-tstrickland) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:17]: I must be doing something wrong then, as a pull on this branch, with a make clean dev results in something that won't boot.
`Caused by: java.io.FileNotFoundException: api-enumerations/constants.yml (No such file or directory)
	at java.io.FileInputStream.open0(Native Method) ~[na:1.8.0_252]
	at java.io.FileInputStream.open(FileInputStream.java:195) ~[na:1.8.0_252]
	at java.io.FileInputStream.<init>(FileInputStream.java:138) ~[na:1.8.0_252]
	at uk.gov.companieshouse.web.companydocuments.util.YamlResourceMapper.fetchYamlFromFile(YamlResourceMapper.java:16) ~[classes!/:unversioned]`

Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:17]: I've adapted the `Makefile` to suit, try now

Comment (ch-tstrickland) [reply src/main/java/uk/gov/companieshouse/web/companydocuments/util/EnumerationConstantsMapper.java:17]: Fixed by your last commit to the Makefile.

**Review by ch-tstrickland — APPROVED** [review]
